Android - Clean Up

KelinciFX opened the pull request:
- AndroidManifest.xml.template ==> For better understanding each elements, better looking, and should prefend messing up with added chunk attributes (at least, for me). Also, some R.attr for configChanges was added.
- build.gradle.template ==> In the same spirit of mentioned above.
- detect.py ==> Reverted android-14 to android-15 as 15 close to 4.+ and/or 5+ while still supporting android 2.3.3 (Gingerbread-API10) quite well.

KelinciFX committed:
Some Cleaning Up on :
- AndroidManifest.xml.template ==> For better understanding each elements, better looking, and should prefend messing up with added chunk attributes. Also, some R.attr for configChanges was added.
- build.gradle.template ==> In the same spirit of mentioned above.
- detect.py ==> Reverted android-14 to android-15 as 15 close to 4.+ and/or 5+ while still supporting android 2.3.3 (Gingerbread-API10) quite well.

akien-mga commented on platform/android/build.gradle.template:
```diff
@@ -36,31 +36,34 @@ android {
 			resources.srcDirs = [
-				'res'
-				$$GRADLE_RES_DIRS$$  
+				'src'
```
I think this was really meant to be `res`.

KelinciFX commented:
I'm pretty sure from the very beginning that it should be "src", instead of res. I'd read a book about Android NDK development with Gradle. It also made android package more organized (SS below).

![res vs src](https://cloud.githubusercontent.com/assets/12244526/17164213/4515770c-53f0-11e6-97bf-f01df163763d.png)

volzhs commented:
then, are all drawable\*, value\*, layout folders under ``res`` right?

KelinciFX commented:
They should be there, but some apparently not. I rare touched those files when working with Godot, so It's out of my sync. I guess, it should not be hard to make them included to "res" folder when exporting.

KelinciFX committed:
Proper Handling for "res" folder. Drawable, layout, and values Folders are now included in Apk packaging.

KelinciFX commented:
It seems this pr is now complete.

:-)

volzhs commented on platform/android/build.gradle.template:
```diff
@@ -35,6 +35,9 @@ android {
 		exclude 'META-INF/NOTICE'
+		exclude 'res/drawable/icon.png'
+		exclude 'res/layout/downloading_expansion.xml'
+		exclude 'res/layout/status_bar_ongoing_event_progress_bar.xml'
```
why those files are excluded?

KelinciFX replied:
They are actually included. Gradle made two of each those files in different location, and then complain about "duplication" when exporting. By doing exlude as suggested by Gradle itself, the problem resolved. And of course, those three files are still in their respective folders.

KelinciFX replied:
Update classpath build tools to 2.1.2 to match the latest version of Android Studio as of this commit.

![buildtools_212](https://cloud.githubusercontent.com/assets/12244526/17388744/534386dc-5a28-11e6-91e1-8b5d69335322.png)

:-)

RandomShaper commented:
Could you elaborate a bit more the reason of reverting the _android-14_ NDK platform to _android-15_?

KelinciFX commented:
My reasons : 
1) android-14 sdk is obsolete.
2) android-15 sdk is not obsolete and supports android-10. So, it should work on android-14 as well (Afaiu API-15 was mostly meant bugfixes for API-14).

:-)

RandomShaper commented:
As per https://developer.android.com/ndk/guides/stable_apis.html the NDK
platform 14 is still a viable option.

The obsolete platform 14 is the SDK one. But they go hand in hand so
deciding to target 15 could make sense. Anyway, the minSdkVersion in the
final manifest should reflect that.

Bu the way, I'm now a bit confused, which is the lowest Android version
Godot games are expected to run on? In the manifest template I see 10, but
the Gradle buildscript specifies 14/15.

I think we cannot reliably support 2.3 if we link against a newer NDK (for
the SDK that's no problem as long as the relevant API checks are done at
runtime).

El 3/8/2016 7:38 p. m., "Kelinci Senja" <notifications@github.com> escribió:

> My reasons :
> 1) android-14 sdk is obsolete.
> 2) android-15 sdk is not obsolete and supports android-10. So, it should
> work on android-14 as well (Afaiu API-15 was mostly meant bugfixes for
> API-14).
>
> :-)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/5915#issuecomment-237298645>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCttOPciknHg76k29eP6Rr3xkG4tJKks5qcNEigaJpZM4JU0cA>
> .
>

reduz commented:
Forget about 2.3, it's way too hacky to get stuff to work on it

On Wed, Aug 3, 2016 at 7:45 PM, Pedro J. Estébanez <notifications@github.com
> wrote:

> As per https://developer.android.com/ndk/guides/stable_apis.html the NDK
> platform 14 is still a viable option.
>
> The obsolete platform 14 is the SDK one. But they go hand in hand so
> deciding to target 15 could make sense. Anyway, the minSdkVersion in the
> final manifest should reflect that.
>
> Bu the way, I'm now a bit confused, which is the lowest Android version
> Godot games are expected to run on? In the manifest template I see 10, but
> the Gradle buildscript specifies 14/15.
>
> I think we cannot reliably support 2.3 if we link against a newer NDK (for
> the SDK that's no problem as long as the relevant API checks are done at
> runtime).
>
> El 3/8/2016 7:38 p. m., "Kelinci Senja" <notifications@github.com>
> escribió:
>
> > My reasons :
> > 1) android-14 sdk is obsolete.
> > 2) android-15 sdk is not obsolete and supports android-10. So, it should
> > work on android-14 as well (Afaiu API-15 was mostly meant bugfixes for
> > API-14).
> >
> > :-)
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/godotengine/godot/pull/5915#issuecomment-237298645>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ALQCttOPciknHg76k29eP6Rr3xkG4tJKks5qcNEigaJpZM4JU0cA
> >
> > .
>
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/5915#issuecomment-237398359>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2529i6IYlbD4nvz9q_6Y9KOTxzovks5qcRn2gaJpZM4JU0cA>
> .
>

KelinciFX commented:
I'm aware that AndroidManifest.xml.template should follow build.gradle.template. But, I'm driven to not doing that by nature because I always build Godot for Android 2.3.3 as minimum API supported. Until now, Godot always run quite well on my old device running Gingerbread using custom build. I though a decision to cut API-10 was not a good strategy for Godot existent.

Yes, 2.3.3 is very old one, but Godot runs quite well on it. Jetpaca is playable there. If there are any issue/problem when playing Godot game on Android 2.3.3 devices (but so far, I never encounter any issue), it's of course not Godot fault. We can't blame Godot for that as 2.3.3 has its own limitations (for being a very old one). It's OK to not supporting 2.3.3, but allowing Godot to be played by millions of users is indeed a good idea.

So, I always recommend Godot to be allowed to run on Android 2.3.3 devices for the following reasons:
1) Android 2.3.3-2.3.7 takes >1.7% of hundreds of millions mobile devices in more than 190 countries.
2) Godot still uses OpenGLES 2.0 only (not yet 3.0, right?), while on Android itself, OpenGLES 3.0 start to be supported by Google team for API-18 (as minimum).
3) Godot runs quite well on Android 2.3.3.
4) I read many Android developers a little bit shocked once they can't test Godot on their Android 2.3.3.
Not because Godot can't run on it, but was not built for the OS in mind.
5) Many OSS and/or Proprietary App/Game making softwares still give Android 2.3.3 a chance to play app/game made with it. The market still considered as somewhat "huge". It means, they reach more and more users. I'm sure it's a good idea to join the market, grab more users from it, and show the existence of Godot there.
6) I don't say that Godot MUST/SHOULD support Android 2.3.3. I'm just saying
- "Make Android 2.3.3 Users Allowed to Enjoy Godot Game Again"
- "Make Godot Grabs Even More Users Again"
- "Make Developers Targeting API-10 Smile Again"
7) Google also mentioned that lower API level targeting more devices, have fewer features available. So, users can't blame Godot Game if they encounter any issue. Perhaps, this is the reason why I often read something like this "The game apparently run on an old Android OS powered device. It may be playable, but be warned that some features may not work"

![selecting_api-10_as_a_minimum_api_targeted](https://cloud.githubusercontent.com/assets/12244526/17386353/42d492c8-5a14-11e6-9ee7-5c8bde68f704.png)

:-)

KelinciFX committed:
Even More Loves for Android Things :

- Updating Gradle stuff.
- Give Android 2.3.3 Users (and Developers) a second chance to enjoy Godot powered Apps/Games.
- My "secret" in compiling Godot for android is now completely revealed. No hidden gems.

:-)

KelinciFX commented on platform/android/build.gradle.template:
```diff
@@ -3,7 +3,7 @@ buildscript {
 	}
 	dependencies {
-		classpath 'com.android.tools.build:gradle:2.1.0'
+		classpath 'com.android.tools.build:gradle:2.1.2'
```
Update classpath build tools to 2.1.2 to match the latest version of Android Studio as of this commit.

![buildtools_212](https://cloud.githubusercontent.com/assets/12244526/17388744/534386dc-5a28-11e6-91e1-8b5d69335322.png)

:-)

KelinciFX commented on platform/android/build.gradle.template:
```diff
@@ -35,32 +35,38 @@ android {
+
 	defaultConfig {
-		minSdkVersion 14
+		minSdkVersion 10
```
Give Android 2.3.3 Users a second chance to enjoy Godot powered Apps/Games. This doesn't mean Godot MUST/SHOULD support API-10. AFAIU Godot runs quite well on Gingerbread. By showing All Godot capabilities, I hope it could attract even more Android developers to use Godot.

This could be changed eventually when Godot can't be played on Android API-10 powered devices.

:-)

KelinciFX commented:
Gradle wrapper version on Upstream is not what it should. The intended Gradle version is 2.10, while the wrapper comes from version 1.6. But, it's now repaired by last commit on Android_CleanUp. Let's take a look at the pict below : 

![gradlewrapper_upstream_vs_androidcleanup](https://cloud.githubusercontent.com/assets/12244526/17389670/888ff0ca-5a30-11e6-8d65-edb3b7e52cba.png)

:-)

volzhs commented:
There must be some reason why Godot dev team decided to support Android 4.0+ not 2.3+.
I like supporting more devices, but should respect their decision.

volzhs commented:
If someone really wants to support Android 2.3+, he can compile their own templates for it.

KelinciFX commented:
Sure. People are always welcomed to compile their own templates. Though, not all people would want to do it. But, it's simply not about that, instead already stated on my previous comments above. About respecting Godot team decision, I'm pretty sure this PR has never been planned and submitted without spirit in respecting other Godot developers (core team, contributor, etc). It's OK to not believe it. If I may... To be honest, I felt that I am not respected.

As I found that my PR just made any troubles instead of making Godot better (the battery of my PR) and to avoid many other troubles, I decided to close it. I apologize to everyone for my bads. Thank you...

:-)

volzhs commented:
Personally, I appreciate your cleaning job, like resource structure.
There are just some arguments for detail things.

RandomShaper commented:
Android 2.3.x has a share of 1.7% according to [this](https://developer.android.com/about/dashboards/index.html).

One approach could be striving to keep support for it to avoid exluding those users, but eventually (and probably in a short time) it would be very difficult (if not impossible) to keep it so it's not worth. Furthermore, probably a lot of those devices will ony be able to run a subset of modern games made with Godot because of CPU/GPU limitations.

RandomShaper commented:
@KelinciFX, I trylu feel your pain regarding not being "respected". Please don't take it that way.

I know it's prerry disappointing to make some work with your best intention and trying to explain your rationale, which takes time and energy to see that it gets refused. It has happened to myself several times (not only with Godot).

Please don't be discoraged by this. You probably have still a lot of good contributions to make in the future and it's also got a value in itself trying things, joining the conversation and thus learning on that way.

akien-mga commented:
Note that this PR hasn't been refused, there were just comments on some parts of it. Other parts are fine, and if the PR had been split in several commits, some of them could likely have been merged already. Others need discussion, that's the normal process and it shouldn't be seen as a lack of respect; we strive for quality, so we review things before we merge them.
